vpr/src/place/centroid_move_generator.cpp

```diff
@@ -5,23 +5,155 @@
 #include "place_constraints.h"
 #include "move_utils.h"
 
-e_create_move CentroidMoveGenerator::propose_move(t_pl_blocks_to_be_moved& blocks_affected, t_propose_action& proposed_action, float rlim, const t_placer_opts& placer_opts, const PlacerCriticalities* /*criticalities*/) {
-    //Find a movable block based on blk_type
+#include <queue>
+
+
+// Static member variable definitions
+vtr::vector<NocGroupId, std::vector<ClusterBlockId>> CentroidMoveGenerator::noc_group_clusters_;
+vtr::vector<NocGroupId, std::vector<ClusterBlockId>> CentroidMoveGenerator::noc_group_routers_;
+vtr::vector<ClusterBlockId, NocGroupId> CentroidMoveGenerator::cluster_to_noc_grp_;
+std::map<ClusterBlockId, NocGroupId> CentroidMoveGenerator::noc_router_to_noc_group_;
+
+
+CentroidMoveGenerator::CentroidMoveGenerator()
+    : noc_attraction_w_(0.0f)
+    , noc_attraction_enabled_(false) {}
+
+CentroidMoveGenerator::CentroidMoveGenerator(float noc_attraction_weight, size_t high_fanout_net)
+    : noc_attraction_w_(noc_attraction_weight)
+    , noc_attraction_enabled_(true) {
+    VTR_ASSERT(noc_attraction_weight > 0.0 && noc_attraction_weight <= 1.0);
+
+    const auto& cluster_ctx = g_vpr_ctx.clustering();
+    const auto& noc_ctx = g_vpr_ctx.noc();
+
+    // check if static member variables are already initialized
+    if (!noc_group_clusters_.empty() && !noc_group_routers_.empty() &&
+        !cluster_to_noc_grp_.empty() && !noc_router_to_noc_group_.empty()) {
+        return;
+    }
+
+    noc_group_clusters_.clear();
```

Comment:
I think this long constructor code would be better moved to a helper function with a descriptive name.